### PR TITLE
[Azure Pipelines] Run affected tests on Safari Technology Preview

### PR DESCRIPTION
**Pull request by foolip**

### CODE REVIEW

File: tools/wpt/run.py
Comment (jgraham) [inline tools/wpt/run.py]: `if kwargs.pop("affected")`

File: tools/wpt/run.py
Comment (jgraham) [inline tools/wpt/run.py]: Remove this with the change to `pop`

File: tools/wpt/run.py
Comment (jgraham) [inline tools/wpt/run.py]: I was thinking it should be the union of the test list and the affected tests. I think that makes some sense for interactive use.

File: tools/wpt/testfiles.py
Comment (jgraham) [inline tools/wpt/testfiles.py]: This isn't quite right; if revish is a range it will return the files changed in that range rather than since that range.

**Review by jgraham — DISMISSED** [review]

File: tools/wpt/run.py
Comment (foolip) [reply tools/wpt/run.py]: Rather `revish = kwargs.pop("affected")` and `if revish` I guess?

File: tools/wpt/run.py
Comment (foolip) [reply tools/wpt/run.py]: Yeah, that's good, saves me the trouble of writing a good error message.

File: tools/wpt/testfiles.py
Comment (foolip) [reply tools/wpt/testfiles.py]: Yeah, you're right. It doesn't make sense to have anything but HEAD as the end of the range because affected tests are determined using the checkout (working tree = HEAD + modified + new) but you're still right. I'll revert this and leave it alone, again. What I'd prefer is https://github.com/web-platform-tests/wpt/pull/14093 and changing everything that suggests that revish could be a range to taking about a single revision, and possibly even asserting that it is so even though A..B will happen to work with `git diff`.

**Comment by foolip** [comment]
In order to verify that the "affected tests without changes" job even works before merging this PR, the `--affected` flag has to be merged first. I'll split that into a separate PR when it seems like it will work.

File: tools/wpt/run.py
Comment (foolip) [inline tools/wpt/run.py]: @jgraham this is the main outstanding part that I need advice on. This code with the default logger doesn't print anything or create a wpt_report.json if --log-wptreport is used. Where is the code that decides to run all tests if none are given? I think there, is `--affected` is used, would be the place to terminate in a way that achieves what the TODO asks for.

And how do I make the default logger print things?

Comment (gsnedders) [reply tools/wpt/run.py]: `--log-mach -` should make us get the default output again, at least

Comment (foolip) [reply tools/wpt/run.py]: That's fine for CI, but when people run `./wpt run --affected HEAD^1` or similar locally, I'd like there to be clear if no tests are run, instead of silently exiting.

Comment (jgraham) [reply tools/wpt/run.py]: I think this is a generic problem with the default logger. I think the right solution is to allow log messages until a flag is set in the logger and then turn them off once the run starts (since the point of the logger is to be as non-verbose as possible). I don't know if it's worthwhile to do something special here instead of just fixing the generic problem.

File: .azure-pipelines.yml
Comment (Hexcles) [inline .azure-pipelines.yml]: `tools/ci/azure/tox_pytest.yml` uses `PublishTestResults@2`. Is the difference intentional?

File: tools/wpt/testfiles.py
Comment (Hexcles) [inline tools/wpt/testfiles.py]: Nit: two blank lines before top-level class/function definitions is recommended (by PEP8 I think).

File: tools/wpt/run.py
Comment (Hexcles) [inline tools/wpt/run.py]: This is different from the current behaviour of `check_stability`. Namely,

1. `include_uncommitted=True`: why? I think this won't make a difference on CI, but is this an intentional change for local interactive use?
2. `include_new=True`
3. `ignore_changes` isn't taken into into account: this might be a separate concern (the tests *are* affected, but we are unable to run them all because of resource constraints). Besides, IIUC, although the setup in this PR doesn't run stability checks, it'll still attempt to run all testharness.js tests if `testharness.js` is modified, which will still take too long even w/o `--verify`.

File: tools/wpt/testfiles.py
Comment (Hexcles) [reply tools/wpt/testfiles.py]: The original docstring is confusing (inaccurate?), too. If we fail to summarize the behaviour of `git diff` in various different cases, shall we just expose the implementation detail by explicitly saying `git diff [revish]` is what we do under the hood? I think that'd be less bad.

**Review by Hexcles — COMMENTED** [review]
I very much agree that the `--affected` part of your PR should be split into a separate PR. I can run affected tests without the patch using your split out PR on Taskcluster, which will not only help to verify that part of your PR (with fewer variables) but also reduce duplicated code in my work towards https://github.com/web-platform-tests/wpt.fyi/issues/708.

File: .azure-pipelines.yml
Comment (foolip) [reply .azure-pipelines.yml]: Yeah, different tasks are versioned independently.

File: tools/wpt/run.py
Comment (foolip) [reply tools/wpt/run.py]: Yes, I the `=True` args are for local use, although I didn't test them.

testharness.js is excluded by the affected test logic so modifying that actually doesn't cause all those tests to be run. idlharness.js however is not excluded.

File: tools/wpt/testfiles.py
Comment (foolip) [reply tools/wpt/testfiles.py]: Any idea why flake8 doesn't complain about this?

Comment (gsnedders) [reply tools/wpt/testfiles.py]: Because we disabled that, because @jgraham objected to it, viewing it as unnecessary noise.

File: .azure-pipelines.yml
Comment (Hexcles) [inline .azure-pipelines.yml]: According to the [doc](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/utility/publish-build-artifacts?view=vsts), we can explicitly specify `pathtoPublish` to a *file* instead of the default directory. Can we try that?

Comment (foolip) [reply .azure-pipelines.yml]: Sure, I'll try and see what happens

File: .azure-pipelines.yml
Comment (lukebjerring) [inline .azure-pipelines.yml]:
```suggestion
    displayName: 'Publish results'
    pathtoPublish: '$(Build.ArtifactStagingDirectory)/wpt_report.json'
    artifactName: 'report'
```

File: .azure-pipelines.yml
Comment (Hexcles) [inline .azure-pipelines.yml]: I can run this command just fine locally (of course, not using Safari), but on Azure it's complaining about invalid arguments:

```
$ no_proxy='*' ./wpt run --yes --no-pause --no-fail-on-unexpected --no-restart-on-unexpected --affected 'HEAD@{1}' --log-wptreport /Users/vsts/agent/2.142.1/work/1/a/wpt_report.json --channel preview safari
wpt run: error: unrecognized arguments: --affected safari
```

I don't really understand why this fails (while the "with change" counterpart succeeds).

Comment (foolip) [reply .azure-pipelines.yml]: That's because the argument didn't exist before this PR :) Will work when the argument is landed on master first.

**Comment by Hexcles** [comment]
Lol of course, somehow "without changes" didn't cross my mind :)

**Comment by foolip** [comment]
This seems to work now. One weird thing is that when no tests are affected, an empty file is published as the report, but IIUC that's also the case for Taskcluster. @Hexcles @lukebjerring are either of you tracking a fix for that?

The other issue is that the manifest is generated from scratch, so the job is not as fast as it could be. Making it fast actually does matter, as the results show up on wpt.fyi and being able to cycle in tens of seconds or at least less than one minute seems both doable and very valuable. Adding 3+ minutes to that means one has to switch context for each try.

**Comment by lukebjerring** [comment]
Yah I noticed the artifacts get created either way. The receiver handles empty files cleanly though, so it doesn't break anything (see https://github.com/web-platform-tests/wpt.fyi/pull/876)

**Review by jgraham — APPROVED** [review]

**Comment by jgraham** [comment]
I'm not sure why the manifest isn't being downloaded; running `wpt manifest` should try to download it?

**Comment by foolip** [comment]
@jgraham I can add a `./wpt manifest` step first, and see if that results in an incremental update in `./wpt run`. I think the code path taken there just doesn't have the "download or generate" bits? Is there an issue somewhere about consolidating the entry points to ensuring manifest existence?

**Comment by jgraham** [comment]
So ` - template: tools/ci/azure/update_manifest.yml` should run it; I missed that this job doesn't call that. The run frontend should end up doing the same thing, but it's possibly more broken there? Anyway I think just adding the other template to the steps will fix this?

**Review by Hexcles — CHANGES_REQUESTED** [review]

**Comment by Hexcles** [comment]
Confirmed that it is WAI: https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=1862&view=results

The artifacts also look good. Woohoo!

**Review by Hexcles — APPROVED** [review]
LGTM

**Comment by foolip** [comment]
Thanks everyone! Merging now.

**Comment by foolip** [comment]
Oops, blocked by https://github.com/web-platform-tests/wpt/issues/14448.

**Comment by foolip** [comment]
Now fixed, merging!

**Comment by lukebjerring** [comment]
@foolip - for Azure it's fine, but not for Taskcluster - we should open an issue for a report being produced when no tests are affected, as this causes permanently-pending wpt.fyi checks.

**Pull request merged**
